zrml/orderbook/src/lib.rs

```diff
@@ -386,34 +411,52 @@ mod pallet {
             let order_id = <NextOrderId<T>>::get();
             let next_order_id = order_id.checked_add(1).ok_or(ArithmeticError::Overflow)?;
 
-            let order = Order {
-                market_id,
-                side: side.clone(),
-                maker: who.clone(),
-                outcome_asset,
-                base_asset,
-                outcome_asset_amount,
-                base_asset_amount,
-            };
+            let outcome_asset_amount_minus_fees = outcome_asset_amount
+                .saturating_sub(T::ExternalFees::get_fee(market_id, outcome_asset_amount));
 
-            match side {
+            let order = match side {
                 OrderSide::Bid => {
                     T::AssetManager::reserve_named(
                         &Self::reserve_id(),
                         base_asset,
                         &who,
                         base_asset_amount,
                     )?;
+
+                    Order {
+                        market_id,
+                        side: side.clone(),
+                        maker: who.clone(),
+                        outcome_asset,
+                        base_asset,
+                        // maker requests outcome asset amount minus fees from taker(s)
+                        outcome_asset_amount: outcome_asset_amount_minus_fees,
+                        // maker gives full base asset amount (no fees charged yet)
+                        // the taker(s) get this and immediately fees are charged
+                        base_asset_amount,
+                    }
                 }
                 OrderSide::Ask => {
                     T::AssetManager::reserve_named(
                         &Self::reserve_id(),
                         outcome_asset,
                         &who,
-                        outcome_asset_amount,
+                        outcome_asset_amount_minus_fees,
                     )?;
+
+                    Order {
+                        market_id,
+                        side: side.clone(),
+                        maker: who.clone(),
+                        outcome_asset,
+                        base_asset,
+                        // only outcome_asset_amount_minus_fees left to repatriate
+                        outcome_asset_amount: outcome_asset_amount_minus_fees,
+                        // maker requests full base asset amount (no fees charged yet) from taker(s)
+                        base_asset_amount,
+                    }
                 }
-            }
+            };
```

Comment:
Seems to me like this isn't what we want.

1. Alice wants to sell 100 units of A. But for some reason, she's only selling 99. The last unit just sits there on here account. Why?
2. If we calculate the price as `base_asset_amount / outcome_asset_amount`, then the price will be wrong (too high).

I suggest to revert these changes.

Reply:
I've been passionately waiting for this discussion :-) thanks for bringing this to our focus now!

The reasoning behind this was that both parties, taker and maker should pay their fair share of the trade.

For this to be equal, we charge the fee from the `base_asset_amount` in the `fill_order` function, resulting in fees for the one who gets the base asset. However, the one who gets the outcome asset should also pay their fair share (creator fees), resulting in the same percentage share as charged for the base asset, but now for the outcome asset. 

Do I missing something or is my calculation right? But I totally agree, it's confusing, because the maker did specify an exact `outcome_asset_amount` and actually is charged with less. So I would love to avoid this confusing point. So, I am open for suggestions here!

Reply:
> If we calculate the price as base_asset_amount / outcome_asset_amount, then the price will be wrong (too high).

The fees are always taken upfront, so the price shouldn't be changed, because we basically have:

(base_asset_amount - fees) / (outcome_asset_amount - fees)

Reply:
https://github.com/zeitgeistpm/zeitgeist/pull/1167/commits/56679b768b5a99727bf9d139ca74de500a8a360d